File: Algorithms Advanced/others/GraphAlgorithms/Python/Bellman-Ford.py
def bellman_ford(start, nodes, edges):
    dist = [float('inf')] * nodes
    dist[start] = 0

    for _ in range(nodes - 1):
        for u, v, w in edges:
            if dist[u] != float('inf') and dist[u] + w < dist[v]:
                dist[v] = dist[u] + w

    for u, v, w in edges:
        if dist[u] != float('inf') and dist[u] + w < dist[v]:
            print("Negative weight cycle detected")
            return

    for i in range(nodes):
        print(f"Distance to {i}: {dist[i]}")

# Example
nodes = 5
edges = [(0, 1, -1), (0, 2, 4), (1, 2, 3), (1, 3, 2), (1, 4, 2), (3, 2, 5), (3, 1, 1), (4, 3, -3)]
bellman_ford(0, nodes, edges)

File: Algorithms Advanced/others/GraphAlgorithms/C++/Bellman-Ford.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

void bellmanFord(int start, int nodes, vector<vector<int>> &edges) 
{
    vector<int> dist(nodes, INT_MAX);
    dist[start] = 0;

    for (int i = 0; i < nodes - 1; ++i) 
    {
        for (auto edge : edges) 
        {
            int u = edge[0], v = edge[1], w = edge[2];
            if (dist[u] != INT_MAX && dist[u] + w < dist[v]) 
            {
                dist[v] = dist[u] + w;
            }
        }
    }

    for (auto edge : edges) 
    {
        int u = edge[0], v = edge[1], w = edge[2];
        if (dist[u] != INT_MAX && dist[u] + w < dist[v]) 
        {
            cout << "Negative weight cycle detected\n";
            return;
        }
    }

    for (int i = 0; i < nodes; i++) 
    {
        cout << "Distance to " << i << ": " << dist[i] << endl;
    }
}

int main() 
{
    int nodes = 5;
    vector<vector<int>> edges = {{0, 1, -1}, {0, 2, 4}, {1, 2, 3}, {1, 3, 2}, {1, 4, 2}, {3, 2, 5}, {3, 1, 1}, {4, 3, -3}};
    bellmanFord(0, nodes, edges);
    return 0;
}
